src/libs/kata-sys-util/src/fs.rs

```diff
@@ -0,0 +1,234 @@
+// Copyright (c) 2019-2021 Alibaba Cloud
+// Copyright (c) 2019-2021 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::ffi::OsString;
+use std::fs::{self, File};
+use std::io::{Error, Result};
+use std::os::unix::io::AsRawFd;
+use std::path::{Path, PathBuf};
+use std::process::Command;
+
+use crate::{eother, sl};
+
+// from linux.git/fs/fuse/inode.c: #define FUSE_SUPER_MAGIC 0x65735546
+const FUSE_SUPER_MAGIC: u32 = 0x65735546;
+
+/// Get bundle path (current working directory).
+pub fn get_bundle_path() -> Result<PathBuf> {
+    std::env::current_dir()
+}
+
+/// Get the basename of the canonicalized path
+pub fn get_base_name<P: AsRef<Path>>(src: P) -> Result<OsString> {
+    let s = src.as_ref().canonicalize()?;
+    s.file_name().map(|v| v.to_os_string()).ok_or_else(|| {
+        eother!(
+            "failed to get base name of path {}",
+            src.as_ref().to_string_lossy()
+        )
+    })
+}
+
+/// Check whether `path` is on a fuse filesystem.
+pub fn is_fuse_fs<P: AsRef<Path>>(path: P) -> bool {
+    if let Ok(st) = nix::sys::statfs::statfs(path.as_ref()) {
+        if st.filesystem_type().0 == FUSE_SUPER_MAGIC as i64 {
+            return true;
+        }
+    }
+    false
+}
+
+/// Check whether `path` is on a overlay filesystem.
+pub fn is_overlay_fs<P: AsRef<Path>>(path: P) -> bool {
+    if let Ok(st) = nix::sys::statfs::statfs(path.as_ref()) {
+        if st.filesystem_type() == nix::sys::statfs::OVERLAYFS_SUPER_MAGIC {
+            return true;
+        }
+    }
+    false
+}
+
+/// Check whether the given path is a symlink.
+pub fn is_symlink<P: AsRef<Path>>(path: P) -> std::io::Result<bool> {
+    let path = path.as_ref();
+    let meta = fs::symlink_metadata(path)?;
+
+    Ok(meta.file_type().is_symlink())
```

Comment:
This "works", but doesn't handle broken symlinks, which might lead to problems in other parts of the code?

Reply:
This is mainly used to detect that a path is symlink, so it should be reject. We are working on a `safe-path` crate, which aims to harden path resolving related code. We plan to open PR for `safe-path` once this PR get merged.